Add tct-libtpms

joholl opened the pull request:
Add `tcti-libtpms`. As of now, localities and pollhandles are not implemented. While the TPM state is unique to the tcti context, it cannot be shared across processes.

AndreasFuchsSIT commented:
CI:`Da fehlt der header in den _SOURCES...

```
../../src/tss2-tcti/tcti-libtpms.c:18:10: fatal error: 'tcti-libtpms.h' file not found
  CC       src/tss2-sys/api/libtss2_sys_la-Tss2_Sys_GetTctiContext.lo
#include "tcti-libtpms.h"
         ^~~~~~~~~~~~~~~~
```

AndreasFuchsSIT commented on Makefile.am:
```diff
@@ -340,7 +340,7 @@ libtss2_tcti_pcap = src/tss2-tcti/libtss2-tcti-pcap.la
 lib_LTLIBRARIES += $(libtss2_tcti_pcap)
 pkgconfig_DATA += lib/tss2-tcti-pcap.pc
-EXTRA_DIST += lib/tss2-tcti-pcap.map # joho TODO enable tcti-pcap for Win (.def, visual studio files
+EXTRA_DIST += lib/tss2-tcti-pcap.map
```
This should probably be a separate commit... ;-)

joholl replied:
Will do

AndreasFuchsSIT commented on configure.ac:
```diff
@@ -203,6 +203,12 @@ AC_ARG_ENABLE([tcti-pcap],
+AC_ARG_ENABLE([tcti-libtpms],
+            [AS_HELP_STRING([--disable-tcti-libtpms],
+                            [don't build the tcti-libtpms module])],,
+            [enable_tcti_libtpms=yes])
```
should we default to "check" and have the test on libtpms determine this ?

joholl replied:
Hm, I'm not sure. We do not check for the simulator binaries either (at least when we decide if we build the tcti modules). I though about emitting a configure warning if libtpms is not installed, but most users wouldn't and shouldn't care about that, anyway. I think I'd rather have it enabled by default... Usually it doesn't hurt and if someone wants it disabled, they can.

AndreasFuchsSIT replied:
My thing is: Do we want to pull in the dependency on libtpms into the build per default. Since we use the headers, we would.
For tcti-mssim/swtpm we don't do so.

AndreasFuchsSIT commented on configure.ac:
```diff
@@ -203,6 +203,12 @@ AC_ARG_ENABLE([tcti-pcap],
+                            [don't build the tcti-libtpms module])],,
+            [enable_tcti_libtpms=yes])
+AM_CONDITIONAL([ENABLE_TCTI_LIBTPMS], [test "x$enable_tcti_libtpms" != xno])
+
```
I guess, libtpms does not have a .pc file ? In that case we should call AC_CHECK_LIB somewhere.

AndreasFuchsSIT replied:
My bad, we're not linking, so it should be AC_CHECK_HEADERS not AC_CHECK_LIB

joholl replied:
Added.

AndreasFuchsSIT commented on man/Tss2_Tcti_Libtpms_Init.3.in:
```diff
@@ -0,0 +1,113 @@
+.sp
+The
+.I conf
+parameter is currently unused.
```
can we use this for `--tpmstate dir` setting ?

joholl replied:
That was exactly what I was thinking about. We could register callbacks to dump state into a file. This is a bit of work, though, because you can either register no callback or all of them... which you then have to implement. This is a topic for another day and another PR ;)

AndreasFuchsSIT replied:
Oh, ok. Maybe you could add this info to the doc, that the state is not persistenced.
I assumed it would just write to $PWD.

joholl replied:
I reworked the PR, including having a state file.

AndreasFuchsSIT commented on src/tss2-tcti/tctildr-dl.c:
```diff
@@ -56,6 +56,10 @@ struct {
         .description = "Access to libtss2-tcti-mssim.so",
     },
+    {
+        .file = "libtss2-tcti-libtpms.so.0",
```
IMHO, we should not add libtpms to the autoloading logic of tctildr.

Otherwise, you'll never detect a badly setup system, which can be kind of surprising. I would want to keep the current behavior where the tss fails if you messed up.

joholl replied:
Agreed. This would lead to the TSS failing in weird ways when there is no real TPM but libtpms installed (_"Why has my tpm2_startup returned success but now I get error 0x100 not initialized"_).

AndreasFuchsSIT reviewed (changes requested):
Very nice work !

Some comments inline.

Also, we need a (unit) test for this to be merged.

williamcroberts commented:
cirrus is fixed (https://github.com/tpm2-software/tpm2-tss/pull/2003) so you can just rebase and force push. So that will fix one of the failures here.

tstruk commented on src/tss2-tcti/tcti-libtpms.c:
```diff
@@ -0,0 +1,317 @@
+    if (tcti_libtpms == NULL) {
+        return;
+    }
+
```
Call TPMLIB_Terminate() otherwise something might leak

tstruk commented on src/tss2-tcti/tcti-libtpms.c:
```diff
@@ -0,0 +1,317 @@
+        return;
+    }
+
+    if (tcti_common->state == TCTI_STATE_RECEIVE) {
```
Maybe that should be if (!tcti_libtpms->response_buffer)

joholl replied:
Agreed, that would be more elegant.

tstruk commented on src/tss2-tcti/tcti-libtpms.c:
```diff
@@ -0,0 +1,317 @@
+    }
+
+    tcti_libtpms->TPMLIB_MainInit = dlsym(tcti_libtpms->libtpms, "TPMLIB_MainInit");
+    if (tcti_libtpms->TPMLIB_ChooseTPMVersion == NULL) {
```
if (tcti_libtpms->TPMLIB_MainInit)

joholl replied:
Done.

tstruk commented on src/tss2-tcti/tcti-libtpms.c:
```diff
@@ -0,0 +1,317 @@
+    }
+
+    tcti_libtpms->TPMLIB_Process = dlsym(tcti_libtpms->libtpms, "TPMLIB_Process");
+    if (tcti_libtpms->TPMLIB_ChooseTPMVersion == NULL) {
```
if (tcti_libtpms->TPMLIB_Process)

tstruk commented on src/tss2-tcti/tcti-libtpms.c:
```diff
@@ -0,0 +1,317 @@
+TSS2_RC
+tcti_libtpms_dl(TSS2_TCTI_LIBTPMS_CONTEXT *tcti_libtpms)
+{
+    tcti_libtpms->libtpms = dlopen("libtpms.so", RTLD_LAZY | RTLD_LOCAL);
```
what if that is called libtpms.so.0?

joholl replied:
I did not think about that. Now `dlopen()` tries `libtpms.so` and `libtpms.so.0`.

joholl commented:
Thank you all for your feedback! I reworked the PR and added state handling (if a path is given for `conf`) and unit tests. 

* The state file is mapped to memory for performance and simpleness
* The libtpms callbacks are mainly stubbed because libtpms kept creating NVChip (which is insufficient for several reasons)
* Since the callbacks don't support passing opaque data blobs, I needed global variables (indexed by thread ids) and therefore pthread mutexes when writing these
* The TCTI should be thread-safe as long as none of the libtpms-tcti instances are shared among multiple threads

tstruk commented on src/tss2-tcti/tcti-libtpms.c:
```diff
@@ -0,0 +1,969 @@
+    LOG_ERROR("Not implemented");
+
+    return TPM_FAIL;
+}
```
So can't we just use NULL for these not implemented callbacks in TPMLIB_RegisterCallbacks()

joholl replied:
Yes, we could. I decided to have them in the code anyway, so if someone is working on that code in the future, they're instantly aware of these callbacks. E.g. if you decide to implement pysical presence tests, you see that there is a callback. If you want, I'll remove the unneeded stubs.

tstruk commented on src/tss2-tcti/tcti-libtpms.c:
```diff
@@ -0,0 +1,969 @@
+
+    tcti_libtpms->state_mmap_len = STATE_MMAP_CHUNK_LEN;
+
+    if (tcti_libtpms->state_path != NULL) {
```
Looks like this function should be a noop() if tcti_libtpms->state_path == NULL so let's check this condition first here and return a SUCCESS if it is. Same for tcti_libtpms_load_state().
This will allow to simplify the code in Tss2_Tcti_Libtpms_Init() where these functions are called.

joholl replied:
Yes, that's also a way to do it. Will do.

tstruk commented on src/tss2-tcti/tcti-libtpms.c:
```diff
@@ -0,0 +1,969 @@
+ * libtpms tcti.
+ */
+static tcti_instance current_libtpms_tctis[64] = {0};
+pthread_mutex_t current_libtpms_tctis_lock;
```
I still don't quite get the idea why do we need to map active tcti instances to threads. Is it because the callbacks are called asynchronously? Couldn't find any info about that in the libtpms docs.
Perhaps some general comment on why this is needed would be good here.

joholl replied:
The problem is this: you cannot implement localities without registering at least the locality callback. The callback headers, however, provide no way for passing down _any_ data. Even the `tpm_number` parameter is hard-coded to `0` on libtpms side. This is a problem if we have multiple libtpms-instances - there is no way to distinguish different contexts from within the callback.

My solution is having a global variable and setting the "currently active" libtpms-context before every libtpms API call and clearing it again afterwards. This is why I have the wrapper macro `LIBTPMS_API_CALL()`. This only works because libtpms does _not_ call the callbacks asynchronously.

Of course you could have multiple threads. So in fact, I track the "currently active" tcti _per thread_.

williamcroberts replied:
Since you need a map from thread id to context to get the proper context on libtpms callbacks, you could avoid the complexity by using thread local storage. Keyword __thread has been supported for a long time, and makes it even easier.

```C
__thread THREAD_CONTEXT state;
```

Or you can use pthread local storage API, see https://linux.die.net/man/3/pthread_setspecific
example here: https://gist.github.com/jbwyme/9479813

Then you won't need to manage a global map.

joholl replied:
Dang it, if only I had known that earlier... I'll rewrite that bit.

I intentionally did not use pthread (except for the mutex) because I wanted to support non-pthread (i.e. syscall) threads. I'm not too sure if that makes a difference, though. Guess I'll use `__thread ` now, anyway.

Learnt something today, thanks!

joholl commented on src/tss2-tcti/tcti-libtpms.c:
```diff
@@ -0,0 +1,969 @@
+static int
+tcti_libtpms_register_get_thread_index(TSS2_TCTI_LIBTPMS_CONTEXT *tcti_libtpms)
+{
+    pid_t thread_id = syscall(__NR_gettid);
```
I had that before, but [it's not part of glibc](https://linux.die.net/man/2/gettid):

>  Glibc does not provide a wrapper for this system call; call it using syscall(2).

joholl commented on src/tss2-tcti/tcti-libtpms.c:
```diff
@@ -0,0 +1,969 @@
+    }
+
+    /* set its current tcti to this */
+    current_libtpms_tctis[index].tcti_libtpms = tcti_libtpms;
```
You're right. `tcti_libtpms_register_get_thread_index()` is not supposed to register the tcti context, only the thread. That's a bug.

joholl commented on src/tss2-tcti/tcti-libtpms.c:
```diff
@@ -0,0 +1,969 @@
+    }
+
+    /* did not find thread in the list, for write access we need to lock */
+    pthread_mutex_lock(&current_libtpms_tctis_lock);
```
The part above the mutex (searching for the current thread) is read-only and not problematic.

Everything below the mutex goes through the list anew and does not rely on anything read before (it could have changed since). Therefore I do not think this is racy.

williamcroberts replied:
you need to handle the lock failing.

joholl replied:
Thanks, that's it.

joholl commented on src/tss2-tcti/tcti-libtpms.c:
```diff
@@ -0,0 +1,969 @@
+    }
+
+    *tcti_libtpms = current_libtpms_tctis[index].tcti_libtpms;
+    if (tcti_libtpms == NULL) {
```
Needs to be `if (*tcti_libtpms == NULL)`. Good call!

joholl commented on src/tss2-tcti/tcti-libtpms.c:
```diff
@@ -0,0 +1,969 @@
+
+    *tcti_libtpms = current_libtpms_tctis[index].tcti_libtpms;
+    if (tcti_libtpms == NULL) {
+        LOG_ERROR("No TCTI registered for this thread: %ld", syscall(__NR_gettid));
```
No, we are differentiating threads here, not processes.

joholl commented on src/tss2-tcti/tcti-libtpms.c:
```diff
@@ -0,0 +1,969 @@
+    tcti_libtpms->state_len = 0;
+
+    /* permanent buffer length (big endian) */
+    memcpy(&permanent_buf_len, tcti_libtpms->state_mmap + offset, sizeof(permanent_buf_len));
```
I know. I thought it would be more readable that way, but I can remove it ofc.

williamcroberts commented on lib/tss2-tcti-libtpms.def:
```diff
@@ -0,0 +1,4 @@
+LIBRARY tss2-tcti-mssim
```
shouldn't this be tss2-tcti-libtpms?

joholl replied:
Yep, copy-paste error.

williamcroberts commented on man/tss2-tcti-libtpms.7.in:
```diff
@@ -0,0 +1,15 @@
+with the interface and protocol exposed by the daemon hosting the TPM2
+reference implementation. The interface exposed by this library is defined
+in the \*(lqTSS System Level API and TPM Command Transmission Interface
+Specification\*(rq specification.
```
I thought libtpms was a library version of the backing implementation of a TPM. Does this still require a swtpm or something running?

I guess, thus far in my code review, I am missing the bigger picture of this.

joholl replied:
No, you're right. In fact, swtpm links against libtpms. I'll rewrite this to make it clear.

williamcroberts commented on src/tss2-tcti/tcti-libtpms.c:
```diff
@@ -0,0 +1,969 @@
+}
+
+TSS2_RC
+tcti_libtpms_dl(TSS2_TCTI_LIBTPMS_CONTEXT *tcti_libtpms)
```
why do we need to this, why wouldn't this just be linked with this at build time?

joholl replied:
We need to dlopen libtpms because we want to be able to have different libtpms instances. They have no contexts, but a lot of global variables.

williamcroberts replied:
Got it, interesting way around the problem. For some reason I thought dlopen was RTLD_NOLOAD by default, but I guess not.

williamcroberts commented on src/tss2-tcti/tcti-libtpms.h:
```diff
@@ -0,0 +1,47 @@
+    TPM_RESULT (*TPMLIB_GetState)(enum TPMLIB_StateType, unsigned char **, uint32_t *);
+    TPM_RESULT (*TPMLIB_MainInit)(void);
+    TPM_RESULT (*TPMLIB_Process)(unsigned char **, uint32_t *, uint32_t *, unsigned char *, uint32_t);
+    TPM_RESULT (*TPMLIB_SetState)(enum TPMLIB_StateType, const unsigned char *, uint32_t);
```
I don't understand why we just don't link to libtpms and just call these directly.

williamcroberts reviewed (changes requested):
I think you can simplify the mapping of tcti specific callback data by using thread local data over a global mapping. This way you don't have to deal with locking.

I also don't see why we need to dlopen libtpms, can't we just dynamically link at build time so its recorded in the elf needed section as dumped by ldd?

joholl commented:
Thanks @williamcroberts for the hint with `__thread`. I removed the thread/tcti map and replaced it with a single global variable (thread local storage).

I fear we cannot get rid of the `dlopen()`, though. Libtpms holds global state instead of having contexts.

williamcroberts commented on src/tss2-tcti/tcti-libtpms.c:
```diff
@@ -0,0 +1,855 @@
+    } \
+    current_tcti_libtpms = NULL;
+
+__thread TSS2_TCTI_LIBTPMS_CONTEXT *current_tcti_libtpms;
```
We don't want this extern, static. Plus will get NULL initialization as well. I see NULL checks on it, so we probably want to make sure we initialize it to NULL IIUC.

joholl replied:
I added a `static`, thanks for the hint. Global variables are initialized with `NULL` implicitly, but I added an `= NULL` for clarity.

AndreasFuchsSIT commented:
Sorry for asking that late in the game.
Could we maybe add an additional test that checks the correct coexistence of two tcti instances ?

Basically some
```
Initialize(tcti1);
Initialize(tcti2);
PcrExtend(tcti1, pcr0, somedigest);
PcrRead(tcti2, pcr0, tmp);
cmp(tmp, 00000);
```

joholl commented:
@AndreasFuchsSIT Thanks for the suggestion. What do you have in mind? For a unit test, the pcr functionality is mocked away.

On the integration test side, there are no tcti tests. In this case, one would need to write a test framework for tctis (similar to the `sys-...`, `esys-...` and `fapi-...` frameworks, including helpers for arg parsing, env var handling, test setup/teardown and build script stuff). Is that what you have in mind? In that case, maybe this would be a topic for a separate PR.

AndreasFuchsSIT commented:
Since this is a test that is specific to tcti-libtpms (the default tcti-swtpm does not support multiple tctis) I'd make it a unit test.
My though was to just send the bytebuffers for PCRextend and PCRread to the tcti and check the results.
I'm personally not too worried about the dlopen in a unit test. If that's your concern, we'd have to mock dlopen/sym and the tcti calls of course.

joholl commented:
@AndreasFuchsSIT 

> Since this is a test that is specific to tcti-libtpms (the default tcti-swtpm does not support multiple tctis) I'd make it a unit test.
My though was to just send the bytebuffers for PCRextend and PCRread to the tcti and check the results.

Ok, I added two tests that use two tcti instances. One with statefile, one without statefile.

I mocked away dlopen (like for all the new tests). However, for this test specifically, I un-mocked file io to actually show the states do not interfere. The statefile is deleted at the end of the tests (and at the beginning, if necessary).

AndreasFuchsSIT commented:
@diabonas Could you help with adding libtpms to lgtm ?

diabonas commented:
> @diabonas Could you help with adding libtpms to lgtm ?

Sure, I believe the following changes to `.lgtm.yml` should do the trick:
```diff
diff --git a/.lgtm.yml b/.lgtm.yml
index a886ba14..c54d7768 100644
--- a/.lgtm.yml
+++ b/.lgtm.yml
@@ -1,4 +1,16 @@
 extraction:
   cpp:
     prepare:
-      packages: "autoconf-archive"
+      packages:
+      - autoconf-archive
+      - libssl-dev
+    after_prepare:
+    - cd "$LGTM_WORKSPACE"
+    - mkdir installdir
+    - git clone https://github.com/stefanberger/libtpms.git
+    - cd libtpms
+    - ./bootstrap.sh
+    - ./configure --prefix="$LGTM_WORKSPACE/installdir/usr"
+    - make install
+    - export PKG_CONFIG_PATH="$LGTM_WORKSPACE/installdir/usr/lib/pkgconfig:$PKG_CONFIG_PATH"
+    - export LD_LIBRARY_PATH="$LGTM_WORKSPACE/installdir/usr/lib:$LD_LIBRARY_PATH"
```

joholl commented:
@diabonas Thank you very much. I had to add a `C_INCLUDE_PATH` for the header files. Now, LGTM is green though!

@AndreasFuchsSIT Should be ready to merge now

joholl commented:
@AndreasFuchsSIT ping

AndreasFuchsSIT commented:
One final Question: Will I break the tools and other's CIs if I hit merge here, if they don't have libtpms installed ?

If I see it correctly, we are currently tcti-libtpms enabled unless disabled.
Should we maybe turn this to auto-detection, if neither enable nor disable are defined in configure ?

joholl commented:
@AndreasFuchsSIT No, autodetection is already implemented. The build is only aborted if both `--enable-tcti-libtpms` is given and the header is missing. To be more precise:

* if param `--enable/disable-tcti-libtpms` is given
  * if arg is `--enable-tcti-libtpms`
    * if header `libtpms/tpm_library.h` is not present
      * Error: "library libtpms missing"
* else (param not given)
  * if header `libtpms/tpm_library.h` is present
      * enable `libtpms-tcti`
  * else (header `libtpms/tpm_library.h` is not present)
      * disable `libtpms-tcti` and warn: "library libtpms missing"


Relevant code for reference:
```m4
AC_ARG_ENABLE([tcti-libtpms],
              [AS_HELP_STRING([--disable-tcti-libtpms], [don't build the tcti-libtpms module])],
              [AS_IF([test "x$enable_tcti_libtpms" = "xyes"],   # <-- given, if --enable check header and error if not exists
                     [AC_CHECK_HEADER(libtpms/tpm_library.h, [], [AC_MSG_ERROR([library libtpms missing])])])],
              [AC_CHECK_HEADER(libtpms/tpm_library.h,           # <-- not given, check header, enable if exists, disable and warn otherwise
                               [enable_tcti_libtpms=yes],
                               [AC_MSG_WARN([library libtpms missing])])])
AM_CONDITIONAL([ENABLE_TCTI_LIBTPMS], [test "x$enable_tcti_libtpms" != xno])
```
